### DISCUSSION TITLE
Having a conflict between client-side/server-side components when a state-using context is involved

### DISCUSSION THREAD

**Discussion by rjray** [vercel/next.js#71166] [Help]
### Summary

When using a context-provider that uses the `useState` hook, we can't put it in a non-client-side component. If we convert the component to client-side, then we can't use metadata to set things like title on the page.

### Additional information

We have a fairly-complex application (internal and proprietary, unfortunately). We are using the latest Next.js, and have metadata being set in each `layout.tsx` file. However, we have a custom context-provider that we've created, to provide centralized store/retrieve operations on some of the user data. Currently, this is in a `page.tsx` that is two levels below root (call root "level 0"). But it needs to go up one level so that the context it provides is accessible to the home page (level 1) as well as the level 2 page (and other level 2 pages in the future).

(That might be a bit confusing: the whole application is a centralized web app that will have distinct "sub-applications". The level 0 home page will act as a portal to the hosted sub-apps. This problem is coming from a sub-app, which starts with a home page at level 1. The context element is currently in a page file at level 2, but needs to be on level 1 in such a way that all routes from that point in the tree have access to it.)

We've tried using the `Head` component from `next/head` within the markup created in `page.tsx` (along with `"use client"`) at level 1, but that had no effect on the title. Setting the title with `Metadata` in the level 0 layout file is fine for the portal page, but each sub-app is unable to set the title more specifically.

I read the discussion [here](https://github.com/vercel/next.js/discussions/50872), so I know that the recommended technique is to keep both layout and page files as server components and defer client-side operations to child components. But this would put the context local to just the one route target, I think?

What is the recommended way, in this case?

**Comment by lumirlumir** [comment]
Take a look at Boundaries part of the link below. There might be an answer.

https://www.joshwcomeau.com/react/server-components/#boundaries

It is the recommended way of handling mixed client and server components as I know.

**Reply by rjray** [reply]
This may be the explanation and illustration that I needed, in this case. Pretty sure I've skimmed this article before in the past, but focusing on this section is very helpful. Thanks!
